File: for profiler/lib/radix.py
from lib.greedy import *
import math



class RadixSpline:
    def __init__(self,GreedySpline,bitcount):
        self.bitcount=bitcount
        self.greedyspline=GreedySpline.greedyspline
        self.error=GreedySpline.error
        self.radixspline=[None]*2**self.bitcount
        self.bit_len=self.bit_lenght(self.greedyspline[-1][0])
        # print("bit_len",self.bit_len)
        self.len_greedy=len(self.greedyspline)
        self.not_prefics,self.mod=self.not_prefics_count()
        self.move_right=self.bit_len-self.bitcount-self.not_prefics
        # print("not_prefics",self.not_prefics)
        # print("mod", self.mod)
        # print("move_right",self.move_right)
        self.get_RadixSpline()
        
    def bit_lenght(self,n):
        return(len(bin(n).lstrip('-0b')))
    def not_prefics_count(self):
        n=self.greedyspline[0][0]
        count,pref=0,0
        
        if self.bit_lenght(n)==self.bit_len:
            count=self.bit_len-self.bit_lenght(((1<<self.bit_len)-1)^n)
        for i in range(count):
            pref=(pref<<1)+1
        pref<<=(self.bit_len-count)
        return count,pref
        
        
    
    ##Способ построение радикс сплайна, нужно заменить
    def first_bits(self,n):
        if self.not_prefics:
           n%=self.mod
        #    print(n)
        return n>>self.move_right
    
    def get_RadixSpline(self):
        
        for ind in range(len(self.greedyspline)):
            temp=self.first_bits(self.greedyspline[ind][0])
            # print(temp)
            if not self.radixspline[temp]:
                self.radixspline[temp]=ind
        if  self.radixspline[2**self.bitcount-1]==None:
             self.radixspline[2**self.bitcount-1]=len(self.greedyspline)-1
        for i in range (2**self.bitcount-1,-1,-1):
            if self.radixspline[i]==None:                
                self.radixspline[i]=self.radixspline[i+1]
        return self.radixspline
    
    def bin_search_spline(self,left,right,key):
        while left<=right:
            mid=(left+right)//2
            if key>= self.greedyspline[mid][0] :
                if key<=self.greedyspline[mid+1][0] :
                   return [mid,mid+1]
                else: 
                   left=mid+1
            else:
                right=mid-1


    
        
    def give_answer(self,key):
        first_bit=self.first_bits(key)
        left=self.radixspline[first_bit]-1
        if first_bit==len(self.radixspline)-1:
            right=len(self.greedyspline)-1
        else:
            right=self.radixspline[first_bit+1]
        # left=0
        # right=len(self.greedyspline)-1
        bin_ans=self.bin_search_spline(left,right,key)
        kl,pl=self.greedyspline[bin_ans[0]]
        kr,pr=self.greedyspline[bin_ans[1]]

        return pl+(key-kl)*(pr-pl)//max(1,(kr-kl))
    
    def plot(self, fig):
        
        dataX=[]
        dataY=[]
        for i in  self.radixspline:
            dataX.append(self.greedyspline[i][0])
            dataY.append(self.greedyspline[i][1])
        # plt.plot(dataX,dataY, marker='o', markersize=15)
        fig.add_trace(go.Scatter(x=dataX,y=dataY, name="R",
                                  mode='lines+markers',
                marker=dict(
                    color="red",
                    size=20,
                ),))
        return fig
    

def plot_data(data,fig,error=0 ):
    

    dataX=[]
    errT=[]
    errB=[]
    dataY=[]
    for i,j in  data:
        dataX.append(i)
        dataY.append(j)
        if error:
            errB.append(j-error)
            errT.append(j+error)
    fig.add_trace(go.Scatter(x=dataX,y=dataY, name="data",
                             mode="markers",
                marker=dict(
                    color="blue",
                    size=5,
                ),))
    
    # print(data.__sizeof__())
    if error:
        fig.add_trace(go.Scatter(x=dataX,y=errB, name="-err"))
        fig.add_trace(go.Scatter(x=dataX,y=errT, name="+err"))     
    return fig   

def get_spline(data, ERR,bitcount ,plot=[False,False]): 
    G=GreedySpline(error=ERR)
    G.load_data(data)
    
    # print(len(G.greedyspline)/len(data)) 
    R=RadixSpline(G, bitcount)
    if plot[0]:
        fig=go.Figure()
        
        fig=G.plot(fig)
        fig=R.plot(fig)
        fig=plot_data(data,fig,plot[1]*ERR )
        fig.show()
    return (G,R)

def bin_search(data, key, left, right):
    while left <= right:
        mid = (left + right) // 2  # Находим середину массив
        if data[mid][0] == key:  # Если целевое значение найдено
            return mid
        elif data[mid][0] < key:  # Если целевое значение справа
            left = mid + 1
        else:  # Если целевое значение слева
            right = mid - 1
    return -1  


def get_key(data, spline, key ):
    radix_answer=spline[1].give_answer(key[0])
    left=radix_answer-spline[0].error
    right=radix_answer+spline[0].error
    res=bin_search(data, key[0], max(left,data[0][1]),min(right, data[-1][1]))
    # print(res)
    if res!=key[1]: raise Exception("FAIL", res, key, math.floor(left),math.ceil(right))
    return res

File: cpp_profiler/radix.cpp
#include "radix.hpp"

using namespace std;


RadixSpline::RadixSpline(const GreedySpline& greedySpline, int bitcount)
        : splineType(greedySpline), spline(greedySpline.greedyspline),bitcount(bitcount),  error(greedySpline.error)  {
        radixspline.resize(1 << bitcount, -1);
        bit_len = bit_length(spline.back().first);
        len_greedy = spline.size();
        pair<int, int> result = not_prefix_count();
        not_prefix = result.first;
        mod = result.second;
        move_right = bit_len - bitcount - not_prefix;
        get_RadixSpline();
    }

    void RadixSpline::print_spline() const {
        for (const auto& point : radixspline) {
            std::cout << "(" << point<<")" << std::endl;
        }
    }

    int RadixSpline::bit_length(int n) {
        if (n == 0) return 1;
        return static_cast<int>(log2(n)) + 1;
    }

    pair<int, int> RadixSpline::not_prefix_count() {
        int n = spline[0].first;
        int count = 0, pref = 0;

        if (bit_length(n) == bit_len) {
            count = bit_len - bit_length(((1 << bit_len) - 1) ^ n);
        }
        for (int i = 0; i < count; ++i) {
            pref = (pref << 1) + 1;
        }
        pref <<= (bit_len - count);
        return {count, pref};
    }

    int RadixSpline::first_bits(int n) {
        if (not_prefix) {
            n %= mod;
        }
        return n >> move_right;
    }

    void RadixSpline::get_RadixSpline() {
        int idx=-1;  
        for (size_t ind = 0; ind < spline.size(); ++ind) {
            int temp = first_bits(spline[ind].first);
            if (temp > idx) {
                radixspline[temp] = static_cast<int>(ind);
                idx=temp;
            }
        }

        if (radixspline[(1 << bitcount) - 1] == -1) {
            radixspline[(1 << bitcount) - 1] = static_cast<int>(spline.size()) - 1;
        }
        for (int i = (1 << bitcount) - 2; i >= 0; --i) {
            if (radixspline[i] == -1) {
                radixspline[i] = radixspline[i + 1];
            }
        }
    }


    unsigned long RadixSpline::give_answer(unsigned long key) {
        int first_bit = first_bits(key);

        int left = std::max(radixspline[first_bit] - 1,0);
        int right;
        if (first_bit == static_cast<int>(radixspline.size()) - 1) {
            right = static_cast<int>(spline.size()) - 1;
        } else {
            if (first_bit + 1 < static_cast<int>(radixspline.size())) {
                right = radixspline[first_bit + 1];
            } else {
                right = -1;
            }
        }
        auto bin_ans = splineType.bin_search_spline(key, left, right);
        unsigned long kl = spline[bin_ans[0]].first;
        int pl = spline[bin_ans[0]].second;
        unsigned long kr = spline[bin_ans[1]].first;
        int pr = spline[bin_ans[1]].second;
        
    
        return pl + (key - kl) * (pr - pl) / std::max(1.,((kr - kl)*1.));
    }

// int bin_search(const vector<pair<unsigned long, int>>& data, unsigned long key, int left, int right) {
//     while (left <= right) {
//         int mid = (left + right) / 2;
//         if (data[mid].first == key) {
//             return mid;
//         } else if (data[mid].first < key) {
//             left = mid + 1;
//         } else {
//             right = mid - 1;
//         }
//     }
//     return -1;
// }

int bin_search(const std::vector<std::pair<unsigned long, int>>& data, unsigned long key, int left, int right) {
    auto it = lower_bound(data.begin() + left, data.begin() + right + 1, key,
        [](const std::pair<unsigned long, int>& elem, double key) {
            return elem.first < key;
        });

    if (it != data.end() && it->first == key) {
        return distance(data.begin(), it);
    }
    return -1;
}


int get_key(const vector<pair<unsigned long, int>>& data, RadixSpline& spline, const pair<unsigned long, int>& key) {
    unsigned long radix_answer = spline.give_answer(key.first);
    // std::cout <<  radix_answer<< std::endl;
    int left = static_cast<int>(radix_answer - spline.splineType.error);
    int right = static_cast<int>(radix_answer + spline.splineType.error);

    left = max(left, 0);
    right = min(right, spline.splineType.last_point.second);
    // std::cout <<  spline.splineType.error<< std::endl;

    int res = bin_search(data, key.first, left, right);

    if (res != key.second) {
        throw runtime_error("FAIL: " + to_string(res) + ", " + to_string(radix_answer) +
                            ", " + to_string(left) + ", " + to_string(right));
    }
    return res;
}




// Функция для построения сплайнов
RadixSpline get_spline(const vector<pair<unsigned long, int>>& data, int ERR, int bitcount) {
    GreedySpline G(ERR);
    load_data(G, data);

    RadixSpline R(G, bitcount);
    return R;
}
